File: tps/tp-2024-01-resolvido/tp01/ex02/palindrome_cheker.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define INPUT_BREAK "FIM"

bool isPalindrome(char *string) {
    if (string == NULL)
        return false;

    int i = 0, j = strlen(string) - 1;
    while (i < j) {
        if (string[i] != string[j])
            return false;
        i++;
        j--;
    }
    return true;
}

void processInput() {
    char input[100];

    while (true) {
        fgets(input, sizeof(input), stdin);

        input[strcspn(input, "\n")] = 0;

        if (strcmp(input, INPUT_BREAK) == 0)
            break;

        if (isPalindrome(input))
            printf("SIM\n");
        else
            printf("NAO\n");
    }
}

int main() {
    processInput();
    return 0;
}

File: tps/tp-2024-01-resolvido/tp01/ex01/PalindromeCheker.java
package tp01.ex01;

import java.util.Scanner;
import java.util.function.Consumer;

public class PalindromeCheker {
    static final String inputBreak = "FIM";
    
    static boolean isPalindrome(String string) {
        if(string == null)
            return false;

        for(int i = 0, j = string.length() - 1; i < j; i++, j--) {
            if(string.charAt(i) != string.charAt(j))
                return false;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        processInput(sc, string -> {
            String output = isPalindrome(string) ? "SIM" : "NAO";
            System.out.println(output);
        });
            
        sc.close();
    }

    static void processInput(Scanner sc, Consumer<String> inputOperation) {
        String input;
        while (!(input = sc.nextLine()).equals(inputBreak))
            inputOperation.accept(input);
    }
}